File: CSharp/Graphs/DisjointSet/EvaluateDivision.cs
// You are given an array of variable pairs equations and an array of real numbers values, where equations[i] = [Ai, Bi] and values[i] represent the equation Ai / Bi = values[i]. 
// Each Ai or Bi is a string that represents a single variable.
// You are also given some queries, where queries[j] = [Cj, Dj] represents the jth query where you must find the answer for Cj / Dj = ?.
// Return the answers to all queries. If a single answer cannot be determined, return -1.0.
// Note: The input is always valid. You may assume that evaluating the queries will not result in division by zero and that there is no contradiction.

public class Solution
{
    private class Node
    {
        public Node(string key, double val)
        {
            Key = key;
            Value = val;
        }
        public string Key { get; private set; }
        public double Value { get; private set; }
    }

    public double[] CalcEquation(IList<IList<string>> equations, double[] values, IList<IList<string>> queries)
    {
        CreateGraph(equations, values);
        var result = new double[queries.Count];
        for (int i = 0; i < queries.Count; i++)
        {
            var query = queries[i];
            var from = query[0];
            var to = query[1];
            result[i] = Calculate(from, to, new HashSet<string>());
        }
        return result;
    }

    //DFS
    private double Calculate(string from, string to, HashSet<string> visited)
    {
        if (!(graph.ContainsKey(from) && graph.ContainsKey(to)))
        {
            return -1.0;
        }
        if (from == to)
        {
            return 1.0;
        }

        visited.Add(from);
        var adj = graph[from];
        foreach (var a in adj)
        {
            if (visited.Contains(a.Key))
                continue;

            double ans = Calculate(a.Key, to, visited);
            if (ans != -1.0)
            {
                return ans * a.Value;
            }
        }
        return -1.0;
    }

    private void CreateGraph(IList<IList<string>> equations, double[] values)
    {
        for (int i = 0; i < equations.Count; i++)
        {
            var equation = equations[i];
            var src = equation[0];
            var dest = equation[1];
            var val = values[i];

            Connect(src, dest, val);
            Connect(dest, src, 1 / val);
        }
    }

    private Dictionary<string, List<Node>> graph = new Dictionary<string, List<Node>>();
    void Connect(string from, string to, double value)
    {
        if (!graph.ContainsKey(from))
        {
            graph.Add(from, new List<Node>());
        }

        graph[from].Add(new Node(to, value));
    }

    List<Node> Find(string node)
    {
        if (graph.ContainsKey(node))
            return graph[node];

        return null;
    }
}

File: Applications/Python/Graphs/DisjointSet/EvaluateDivision.py
# You re given n rry of vrible pirs equtions nd n rry of rel numbers vlues, where equtions[i] = [i, Bi] nd vlues[i] represent the eqution i / Bi = vlues[i]. 
# Ech i or Bi is  string tht represents  single vrible.
# You re lso given some queries, where queries[j] = [Cj, Dj] represents the jth query where you must find the nswer for Cj / Dj = ?.
# Return the nswers to ll queries. If  single nswer cnnot be determined, return -1.0.
# Note: The input is lwys vlid. You my ssume tht evluting the queries will not result in division by zero nd tht there is no contrdiction.

class Solution:
    def __init__(self):
        self.graph = {}
        
    class Node:
        def __init__(self, key, value):
            self.key = key
            self.value = value
    
    def calcEquation(self, equations: List[List[str]], values: List[float], queries: List[List[str]]) -> List[float]:
        
        self.createGraph(equations, values)
        
        result = []
        for query in queries:  
            result.append(self.calculate(query[0], query[1], {}))        
            
        return result
        
    def calculate(self, src, to, visited):
        if src not in self.graph and to not in self.graph:
            return -1.0
        
        if src == to:
            return 1.0
        
        visited[src] = True        
        if src not in self.graph: # I do not know why this is needed, c# version does not need this check
            return -1.0
        
        adjList = self.graph[src]        
        
        for adj in adjList:                        
            if adj.key in visited:
                continue
            
            result = self.calculate(adj.key, to, visited)
            if result != -1.0:
                return result * adj.value
        
        return -1.0
    
    def connect(self, src, dest, value):
        if src not in self.graph:
            self.graph[src] = []
            
        self.graph[src].append(self.Node(dest, value))
        
    def createGraph(self, equations, values):
        for e in range(len(equations)):            
            src = equations[e][0]
            dest = equations[e][1]
            value = values[e]
                       
            self.connect(src, dest, value)
            self.connect(dest, src, 1/value)   